File: design-patterns/go/facade/travel/travel-facade.go
// travel-facade.go

package main

import "fmt"

type TravelFacade struct {
	startLat float64
	startLng float64
	endLat   float64
	endLng   float64
	location *Location
	toll     *Toll
	car      *Car
	weather  *Weather
}

func NewTravelFacade(startLat float64, startLng float64, endLat float64, endLng float64) (travelFacade *TravelFacade) {
	travelFacade = &TravelFacade{}
	travelFacade.startLat = startLat
	travelFacade.startLng = startLng
	travelFacade.endLat = endLat
	travelFacade.endLng = endLng
	travelFacade.location = NewLocation(startLat, startLng, endLat, endLng)
	travelFacade.car = NewCar()
	travelFacade.toll = NewToll()
	travelFacade.weather = NewWeather()
	return
}

func (travelFacade *TravelFacade) GetCurrentLocation() *Point {
	return travelFacade.location.GetCurrentLocation()
}

func (travelFacade *TravelFacade) GetLocationInfo(lat float64, lng float64) {
	travelFacade.location.GetLocationDetails(lat, lng)
	travelFacade.weather.GetWeatherInfo(lat, lng)
}

func (travelFacade *TravelFacade) GetRoute() []Point {
	return travelFacade.location.GetFullRoute()
}

func (travelFacade *TravelFacade) GetTotalTollAmount(lat float64, lng float64) {
	fmt.Printf("Total Toll Amount: %v\n", travelFacade.toll.GetTotalToll(lat, lng))
}

func (travelFacade *TravelFacade) OperateCar() {
	fullRoute := travelFacade.location.GetFullRoute()
	travelFacade.car.StartEngine()

	for i := 1; i <= len(fullRoute); i++ {
		nextMove := travelFacade.location.GetNextMove()
		switch nextMove {
		case "straight":
			travelFacade.car.GoStraight()
		case "left":
			travelFacade.car.GoLeft()
		case "right":
			travelFacade.car.GoRight()
		default:
			travelFacade.car.GoBack()
		}
	}
	travelFacade.car.StopEngine()
}

File: design-patterns/ts/facade/travel/travel-facade.ts
// travel-facade.ts

import Car from "./car";
import Direction from "./direction";
import Point from "./point";
import Toll from "./toll";
import Weather from "./weather";


class TravelFacade {
    private startLat: number;
    private startLng: number;

    private endLat: number;
    private endLng: number;

    private direction: Direction;
    private toll: Toll;
    private car: Car;
    private weather: Weather;

    // define constructor
    constructor(startLat: number, startLng: number, endLat: number, endLng: number) {
        this.startLat = startLat;
        this.startLng = startLng;
        this.endLat = endLat;
        this.endLng = endLng;

        // Initialize classes
        this.direction = new Direction(startLat, startLng, endLat, endLng);
        this.car = new Car();
        this.toll = new Toll();
        this.weather = new Weather();
    }

    getRoute(): Point[] {
        return this.direction.getFullRoute();
    }

    getLocationInfo(lat: number, lng: number) {
        this.direction.getLocationDetails(lat, lng);
        this.weather.getWeatherInfo(lat, lng);
    }

    getCurrentLocation(): Point {
        return this.direction.getCurrentLocation();
    }

    operateCar(): void {
        const fullRoute = this.direction.getFullRoute();

        this.car.startEngine();

        for (let point of fullRoute) {
            const nextMove = this.direction.getNextMove();

            switch (nextMove) {
                case "straight": this.car.goStraight(); break;
                case "left": this.car.goLeft(); break;
                case "right": this.car.goRight(); break;
            }
        }

        this.car.stopEngine();
    }

    getTotalTollAmount(lat: number, lng: number) {
        console.log("Total Toll Amount: " + this.toll.getTotalToll(lat, lng));
    }

}

export default TravelFacade;